File: aprendi-data/models/enrollment.py
"""
This module contains the Enrollment model and repo
"""
import logging
import json
from typing import Optional
from pydantic import BaseModel
from models.tables import OrganizationDataTable

logging.basicConfig(level=logging.INFO)
log = logging.getLogger(__name__)
log.info('This is an info message')


class EnrollmentModel(BaseModel):
    """
    This class represents the Enrollment model
    """
    enrollment_id: Optional[str] = None
    org_id: str
    student_id: str
    course_id: str
    term_id: str

    # data
    course_name: str
    teacher_name: str
    period: str


class EnrollmentRepo():
    """
    This class represents the Enrollment repo
    """

    @staticmethod
    def enrollment_pk(org_id: str, student_id: str):
        """
        This method returns the pk for the Enrollment
        """
        return f"ORG#{org_id}#STUDENT#{student_id}#ENROLLMENT"

    @staticmethod
    def enrollment_sk(term_id: str, course_id: str, enrollment_id: str):
        """
        This method returns the sk for the Enrollment
        """
        return f"TERM#{term_id}#COURSE#{course_id}#ENROLLMENT#{enrollment_id}"

    @staticmethod
    def enrollment_course_sk(term_id: str, course_id: str):
        """
        This method returns the sk for the Enrollment
        """
        return f"TERM#{term_id}#COURSE#{course_id}"

    @staticmethod
    def enrollment_term_sk(term_id: str):
        """
        This method returns the sk for the Enrollment
        """
        return f"TERM#{term_id}"

    @classmethod
    def save(cls, model: EnrollmentModel) -> EnrollmentModel:
        """
        This method saves the Enrollment to the database
        """
        pk = cls.enrollment_pk(org_id=model.org_id, student_id=model.student_id)
        sk = cls.enrollment_sk(term_id=model.term_id, course_id=model.course_id, enrollment_id=model.enrollment_id)
        data = {
            "course_name": model.course_name,
            "teacher_name": model.teacher_name,
            "period": model.period
        }
        org = OrganizationDataTable(
            pk=pk,
            sk=sk,
            lsi_sk1=model.enrollment_id,
            data=json.dumps(data))
        org.save()  # Saving to the database
        return model

    @classmethod
    def parse_enrollment(cls, item) -> EnrollmentModel:
        """
        This method takes a dynamodb item and returns a EnrollmentModel
        """
        org_id = item.pk.split("#")[1]
        student_id = item.pk.split("#")[3]
        term_id = item.sk.split("#")[1]
        course_id = item.sk.split("#")[3]
        enrollment_id = item.lsi_sk1
        data = json.loads(item.data)
        course_name = data["course_name"]
        teacher_name = data["teacher_name"]
        period = data["period"]
        return EnrollmentModel(
            org_id=org_id,
            student_id=student_id,
            course_id=course_id,
            enrollment_id=enrollment_id,
            term_id=term_id,
            course_name=course_name,
            teacher_name=teacher_name,
            period=period)

    @classmethod
    def get(cls, org_id: str, student_id: str, term_id: str, course_id: str) -> EnrollmentModel:
        """
        This method gets the Enrollment from the database
        """
        pk = cls.enrollment_pk(org_id=org_id, student_id=student_id)
        sk = cls.enrollment_course_sk(term_id=term_id, course_id=course_id)
        items = OrganizationDataTable.query(pk, OrganizationDataTable.sk.startswith(sk))
        items = [cls.parse_enrollment(item) for item in items]
        return items[0]

    @classmethod
    def get_by_enrollment_id(cls, org_id: str, student_id: str, enrollment_id: str) -> EnrollmentModel:
        """
        This method gets the Enrollment from the database
        """
        pk = cls.enrollment_pk(org_id=org_id, student_id=student_id)
        items = OrganizationDataTable.local_secondary_index1.query(pk, OrganizationDataTable.lsi_sk1 == enrollment_id)
        items = [cls.parse_enrollment(item) for item in items]
        return items[0]

    @classmethod
    def get_all(cls, org_id: str, student_id: str, term_id: str) -> list[EnrollmentModel]:
        """
        This method gets all the Enrollments from the database
        """
        pk = cls.enrollment_pk(org_id=org_id, student_id=student_id)
        sk = cls.enrollment_term_sk(term_id=term_id)
        items = OrganizationDataTable.query(pk, OrganizationDataTable.sk.startswith(sk))
        return [cls.parse_enrollment(item) for item in items]

File: aprendi-api/cmd/aprendi-api/repos/enrollment.go
package repos

import (
	"fmt"
	"strings"
)

// EnrollmentRequest represents the Enrollment request model in Go
type EnrollmentRequest struct {
	OrgID     string          `json:"org_id"`
	StudentID string          `json:"student_id"`
	TermID    string          `json:"term_id"`
	Courses   []CourseRequest `json:"courses"`
}

// CourseRequest represents the Course request model in Go
type CourseRequest struct {
	CourseID string `json:"course_id"`
	Prioty   int    `json:"priority"`
}

// Enrollment represents the Enrollment model in Go
type Enrollment struct {
	EnrollmentID string `json:"enrollment_id,omitempty"`
	OrgID        string `json:"org_id"`
	StudentID    string `json:"student_id"`
	CourseID     string `json:"course_id"`
	TermID       string `json:"term_id"`
	CourseName   string `json:"course_name"`
	TeacherName  string `json:"teacher_name"`
	Period       string `json:"period"`
}

// EnrollmentRepo defines the repository operations for the Enrollment model
type EnrollmentRepo struct {
	baseRepo OrganizationDataRepo
}

// NewEnrollmentRepo initializes a new Enrollment repository
func NewEnrollmentRepo(base OrganizationDataRepo) *EnrollmentRepo {
	return &EnrollmentRepo{
		baseRepo: base,
	}
}

// enrollmentPK generates the primary key for the Enrollment
func (er *EnrollmentRepo) enrollmentPK(orgID, studentID string) string {
	return fmt.Sprintf("ORG#%s#STUDENT#%s#ENROLLMENT", orgID, studentID)
}

// enrollmentSK generates the sort key for the Enrollment
func (er *EnrollmentRepo) enrollmentSK(termID, courseID, enrollmentID string) string {
	return fmt.Sprintf("TERM#%s#COURSE#%s#ENROLLMENT#%s", termID, courseID, enrollmentID)
}

// enrollmentCourseSK generates the course sort key for the Enrollment
func (er *EnrollmentRepo) enrollmentCourseSK(termID, courseID string) string {
	return fmt.Sprintf("TERM#%s#COURSE#%s", termID, courseID)
}

// enrollmentTermSK generates the term sort key for the Enrollment
func (er *EnrollmentRepo) enrollmentTermSK(termID string) string {
	return fmt.Sprintf("TERM#%s", termID)
}

// Save inserts or updates an Enrollment in the database
func (er *EnrollmentRepo) Save(enrollment *Enrollment) error {
	dataMap := make(map[string]string)
	dataMap["course_name"] = enrollment.CourseName
	dataMap["teacher_name"] = enrollment.TeacherName
	dataMap["period"] = enrollment.Period
	data := &OrganizationData{
		PK:     er.enrollmentPK(enrollment.OrgID, enrollment.StudentID),
		SK:     er.enrollmentSK(enrollment.TermID, enrollment.CourseID, enrollment.EnrollmentID),
		LSISK1: &enrollment.EnrollmentID,
		Data:   dataMap,
	}
	return er.baseRepo.Save(data)
}

// ParseEnrollment translates a OrganizationData to an Enrollment
func (er *EnrollmentRepo) ParseEnrollment(data *OrganizationData) (*Enrollment, error) {
	orgID := strings.Split(data.PK, "#")[1]
	studentID := strings.Split(data.PK, "#")[3]
	termID := strings.Split(data.SK, "#")[1]
	courseID := strings.Split(data.SK, "#")[3]
	enrollmentID := data.LSISK1
	return &Enrollment{
		EnrollmentID: *enrollmentID,
		OrgID:        orgID,
		StudentID:    studentID,
		TermID:       termID,
		CourseID:     courseID,
		CourseName:   data.Data["course_name"],
		TeacherName:  data.Data["teacher_name"],
		Period:       data.Data["period"],
	}, nil
}

// Get retrieves an Enrollment based on orgID, studentID, termID, and courseID
func (er *EnrollmentRepo) Get(orgID, studentID, termID, courseID string) (*Enrollment, error) {
	data, err := er.baseRepo.GetByPKSK(er.enrollmentPK(orgID, studentID), er.enrollmentCourseSK(termID, courseID))
	if err != nil || data == nil {
		return nil, err
	}
	return er.ParseEnrollment(data)
}

// GetByEnrollmentID retrieves an Enrollment using the enrollment ID
func (er *EnrollmentRepo) GetByEnrollmentID(orgID, studentID, enrollmentID string) (*Enrollment, error) {
	dataItems, err := er.baseRepo.GetByPK(er.enrollmentPK(orgID, studentID))
	if err != nil {
		return nil, err
	}
	for _, item := range dataItems {
		if item.LSISK1 == &enrollmentID {
			return er.ParseEnrollment(&item)
		}
	}
	return nil, nil
}

// GetAll retrieves all Enrollments for a given orgID, studentID, and termID
func (er *EnrollmentRepo) GetAll(orgID, studentID, termID string) ([]Enrollment, error) {
	dataItems, err := er.baseRepo.GetByPK(er.enrollmentPK(orgID, studentID))
	if err != nil {
		return nil, err
	}

	var enrollments []Enrollment
	for _, item := range dataItems {
		if strings.HasPrefix(item.SK, er.enrollmentTermSK(termID)) {
			enrollment, err := er.ParseEnrollment(&item)
			if err == nil {
				enrollments = append(enrollments, *enrollment)
			}
		}
	}
	return enrollments, nil
}
